### ISSUE TITLE
final class vs interface

### ISSUE THREAD

**Issue by nyamsprod** [delight-im/PHP-Temporal#3]
Looking a the codebase I have to wonder why creating final classes that only contains constants instead of  interfaces.

Something along these lines:

```php
<?php

use Delight\Temporal\Iso8601\Iso8601Date;
use Delight\Temporal\Iso8601\Iso8601DateTime;
use Delight\Temporal\Iso8601\Iso8601Duration;
use Delight\Temporal\Iso8601\Iso8601Time;
use Delight\Temporal\Iso8601\Iso8601Weekday;
use Delight\Temporal\Timestamp\Unix;

/** Immutable representation of date and/or time */
final class Temporal implements 
    Iso8601Date,
    Iso8601DateTime,
    Iso8601DateDuration,
    Iso8601Time,
    Iso8601Weekday,
    Unix
{
... rest of the code here
```

### Why:

- These value will not change so their are valid candidate for being put in an interface
- Someone could easily use your interface to implements them in their classes

### How:

- To avoid BC break you would need to name the interface(s) differently and then copy the current value in the current final class.

### Open Issue

- Should you have one big interface will all the constants of a serie of small interfaces ?

**Comment by ocram** [comment]
Thank you!

I agree that

```
\Delight\Temporal\Iso8601\Iso8601Date
\Delight\Temporal\Iso8601\Iso8601DateTime
\Delight\Temporal\Iso8601\Iso8601Time
\Delight\Temporal\Iso8601\Iso8601Weekday
\Delight\Temporal\Timestamp\Unix
```

(though not `\Delight\Temporal\Iso8601\Iso8601Duration`) could just be changed from `final class` to `interface`, removing their private constructors.

Is this what you were referring to?

However, I don't understand why the `Temporal` class should now have to implement these interfaces that are just wrappers or namespaces for the constants. That change above should be enough, don't you think so?

By the way, this would not break backward compatibility, at least in this form changing the classes to interfaces only.

**Comment by nyamsprod** [comment]
> (though not \Delight\Temporal\Iso8601\Iso8601Duration)

Why ? the only method attached to this class is a static public method meaning

-  it could be attached to the `Duration` class since this method is directly related to quirks in time interval  implementation in PHP.
- it could be converted into a simple namespace function

> However, I don't understand why the Temporal class should now have to implement these interfaces that are just wrappers or namespaces for the constants.

To ease usage and integration. This is well explained here https://github.com/shrikeh/teapot/blob/master/README.md . The `Temporal` class is not required to implement them but It would make life easier for the developper to only consider one entrypoint for your library

**Comment by ocram** [comment]
> Why ? the only method attached to this class is a static public method

Well, it wouldn't make any sense to make that class an interface – just to move its single method to another class or namespace where it fits less. That's what classes are for. Classes are not "bad".

> it could be attached to the `Duration` class since this method is directly related to quirks in time interval  implementation in PHP.

Yes, but it could just as well remain in its current class, which is – without doubt – a better fit, since it specifically affects the ISO 8601 string notation, not durations in general.

> To ease usage and integration. This is well explained here https://github.com/shrikeh/teapot/blob/master/README.md .

I don't think that makes sense. Have never seen that pattern before and can't find any convincing argument on that page, either. Classes should not implement interfaces that they're not really a direct implementation of just to get shorter references. That is what (aliased) imports are for. Anyway, it's a minor point, we can just ignore this part :)

We will make all those classes (except `Iso8601Duration`) interfaces. Thanks again!

**Comment by nyamsprod** [comment]
> Well, it wouldn't make any sense to make that class an interface – just to move its single method to another class or namespace where it fits less. That's what classes are for. Classes are not "bad".

Yes class are not bad but if the only method available is a public static method it means that the method is just a simple function :)

> Yes, but it could just as well remain in its current class, which is – without doubt – a better fit, since it specifically affects the ISO 8601 string notation, not durations in general.

What's affected is not the ISO 8601 notation *but* its interpretation by `DateTime|Immutable` in PHP which is a implementation quirks ... depending on the developers views.

What's true is that your constants are not exposed in your public API so in the end you may choose to keep them where they are now because from a public API stand it does not matter.
